### ISSUE TITLE
Native code order of magnitude slower than translated code on Apple M1

### ISSUE THREAD

**Issue by neurolabusc** [numpy/numpy#17989]
I realize numpy is using experimental compilers for native builds on the M1, and still has some [bugs](https://github.com/numpy/numpy/issues/17964), so it might be premature to discuss optimizations. Perhaps this is a feature request and not a bug. However, one would expect that native ARM code would typically be at least as fast as translated x86-64 code. I noticed that the [nibabel bench_finite_range.py](https://github.com/nipy/nibabel) test is much slower for the native code than translated code. I found translated code (Python 3.8.3, NumPy version 1.19.4) is x10 faster than native code (Python 3.9.1rc1, NumPy version 1.19.4)
### Reproducing code example:

```#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import numpy as np
from numpy.testing import measure
#example where translated code (Python 3.8.3, NumPy version 1.19.4) is x10 faster than native code (Python 3.9.1rc1, NumPy version 1.19.4)
rng = np.random.RandomState(20111001)
img_shape = (128, 128, 64, 10)
repeat = 100
arr = rng.normal(size=img_shape)
mtime = measure('np.max(arr)', repeat)
print('%30s %6.2f' % ('max all finite', mtime))
mtime = measure('np.min(arr)', repeat)
print('%30s %6.2f' % ('min all finite', mtime))
arr[:, :, :, 1] = np.nan
mtime = measure('np.max(arr)', repeat)
print('%30s %6.2f' % ('max all nan', mtime))
mtime = measure('np.min(arr)', repeat)
print('%30s %6.2f' % ('min all nan', mtime))
```

### Performance:

Translated:
```
$ time ./numpy_native_slower_than_translated.py
                max all finite   0.18
                min all finite   0.18
                   max all nan   0.18
                   min all nan   0.19
./numpy_native_slower_than_translated.py  1.32s user 1.28s system 214% cpu 1.213 total
```

Native:
```
$ time ./numpy_native_slower_than_translated.py
                max all finite   1.98
                min all finite   1.99
                   max all nan   1.99
                   min all nan   1.98
./numpy_native_slower_than_translated.py  8.49s user 0.14s system 104% cpu 8.237 total
```

### NumPy/Python version information:

Translated:
 - 1.19.4 3.8.3 (default, May 19 2020, 13:54:14) 
[Clang 10.0.0 ]

Native:
 -  1.19.4 3.9.1rc1 | packaged by conda-forge | (default, Nov 28 2020, 22:21:58) 
[Clang 11.0.0 ]

**Comment by mattip** [comment]
Is there a way to test the HEAD or the numpy 1.20rc1 with native? We have started to use SIMD intrinsics in a way that might make this faster.

**Comment by neurolabusc** [comment]
Happy to help, are there instructions for building on this platform? I have Clang 12.0.0 and gFortran 11.0.0 20201114 (experimental) but not gcc proper. It looks like there are dependencies for bras - do I use brew to install from source? 

```
$pip install numpy==1.20rc1 
Collecting numpy==1.20rc1
  Using cached numpy-1.20.0rc1.zip (7.7 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Building wheels for collected packages: numpy
  Building wheel for numpy (PEP 517) ... error
  ERROR: Command errored out with exit status 1:
   command: /Users/chris/miniforge3/bin/python3.9 /Users/chris/miniforge3/lib/python3.9/site-packages/pip/_vendor/pep517/_in_process.py build_wheel /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpsavxx_s9
       cwd: /private/var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/pip-install-t6cbgk1m/numpy_96013a2b4f96408886c142437efacd24
  Complete output (708 lines):
  Running from numpy source directory.
  numpy/random/_bounded_integers.pxd.in has not changed
  numpy/random/_philox.pyx has not changed
  numpy/random/_bounded_integers.pyx.in has not changed
  numpy/random/_sfc64.pyx has not changed
  numpy/random/_mt19937.pyx has not changed
  numpy/random/bit_generator.pyx has not changed
  Processing numpy/random/_bounded_integers.pyx
  numpy/random/mtrand.pyx has not changed
  numpy/random/_generator.pyx has not changed
  numpy/random/_pcg64.pyx has not changed
  numpy/random/_common.pyx has not changed
  Cythonizing sources
  blas_opt_info:
  blas_mkl_info:
  customize UnixCCompiler
    libraries mkl_rt not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  blis_info:
    libraries blis not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  openblas_info:
    libraries openblas not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  atlas_3_10_blas_threads_info:
  Setting PTATLAS=ATLAS
    libraries tatlas not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  atlas_3_10_blas_info:
    libraries satlas not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  atlas_blas_threads_info:
  Setting PTATLAS=ATLAS
    libraries ptf77blas,ptcblas,atlas not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  atlas_blas_info:
    libraries f77blas,cblas,atlas not found in ['/Users/chris/miniforge3/lib', '/usr/local/lib', '/usr/lib', '/opt/local/lib']
    NOT AVAILABLE
  
  /private/var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/pip-install-t6cbgk1m/numpy_96013a2b4f96408886c142437efacd24/numpy/distutils/system_info.py:1989: UserWarning:
      Optimized (vendor) Blas libraries are not found.
      Falls back to netlib Blas library which has worse performance.
      A better performance should be easily gained by switching
      Blas library.
    if self._calc_info(blas):
  blas_info:
  C compiler: clang -pthread -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/chris/miniforge3/include -arch arm64 -fPIC -O2 -isystem /Users/chris/miniforge3/include -arch arm64
  
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var/folders
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var/folders/4p
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T
  creating /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh
  compile options: '-I/usr/local/include -I/opt/local/include -I/Users/chris/miniforge3/include -c'
  clang: /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/source.c
  /var/folders/4p/cxb1ys0s4wz6dh3rtj22pq8w0000gn/T/tmpyamaythh/source.c:1:10: fatal error: 'cblas.h' file not found
  #include <cblas.h>
           ^~~~~~~~~
  1 error generated.

```

**Comment by mattip** [comment]
`homebrew install openblas` or so?

**Comment by neurolabusc** [comment]
I have been able to install many native libraries with homebrew by installing from source. However, the successful formulas appear to compile with clang, not gcc. It looks to me like openblas demands gcc:

```
$export ARCHFLAGS='-arch arm64'
$brew install -s openblas

Warning: You are running macOS on a arm64 CPU architecture.
We do not provide support for this (yet).
Reinstall Homebrew under Rosetta 2 until we support it.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
unsupported configuration.

==> Downloading https://raw.githubusercontent.com/Homebrew/formula-patches/7baf6e2f/gcc/bigsur.diff
######################################################################## 100.0%
==> Downloading https://ftp.gnu.org/gnu/gcc/gcc-10.2.0/gcc-10.2.0.tar.xz
Already downloaded: /Users/chrisrorden/Library/Caches/Homebrew/downloads/c8d4ca732a98ae691d04472b15de6d9e06a09016499af6ff16c4f55081bfc6b9--gcc-10.2.0.tar.xz

...
Error: You are running macOS on a arm64 CPU architecture.
```

As a Hail Mary, I tried to compile Iain Sandoe's experimental gcc build from source. I could not find any explicit instructions, so used a generic recipe, which did not go well...
```
$git clone https://github.com/iains/gcc-darwin-arm64
$cd gcc-darwin-arm64
$contrib/download_prerequisites
$mkdir build && cd build
$../configure --prefix=/usr/local/gcc-11 \
              --enable-checking=release \
              --enable-languages=c,c++,fortran \
              --disable-multilib \
              --with-sysroot=/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk \
              --program-suffix=-11

$make -j 
...

make[3]: /Users/chrisrorden/src/gcc-darwin-arm64/build/./prev-gcc/xg++: Resource temporarily unavailable
xg++: fatal error: cannot execute ‘/Users/chrisrorden/src/gcc-darwin-arm64/build/./prev-gcc/cc1plus’: vfork: Operation timed out
compilation terminated.
```

**Comment by stweil** [comment]
I compiled OpenBLAS git master with clang and built numpy with it. The result is not faster:

                max all finite   1.97
                min all finite   1.97
                   max all nan   1.97
                   min all nan   1.98
       11.37 real         9.84 user         0.30 sys

A second run starts faster:

        8.28 real        10.09 user         0.03 sys

**Comment by mattip** [comment]
I wonder what SIMD features are exposed in native and in translated mode. Could you try this code with 1.20?
```
python -c "from numpy.core._multiarray_umath import __cpu_features__ as feat; print([f for f in feat if feat[f]])"
```

At some point we should expose a nicer interface to the `__cpu_features__` dictionary of SIMD features detected at runtime.

**Comment by stweil** [comment]
Native ARM with numpy git master:

    % python -c "from numpy.core._multiarray_umath import __cpu_features__ as feat; print([f for f in feat if feat[f]])"
    ['NEON', 'NEON_FP16', 'NEON_VFPV4', 'ASIMD', 'FPHP', 'ASIMDHP']

**Comment by mattip** [comment]
What does the translated processor show? There is SIMD code for avx512f, but none for Neon (yet).

**Comment by stweil** [comment]
That could explain the results. I don't have a translated Python installed. @neurolabusc, could you please test that?

**Comment by stweil** [comment]
> There is SIMD code for avx512f, but none for Neon (yet).

There is a `numpy/core/src/common/simd/neon/` in git master. Which part is missing?

**Comment by mattip** [comment]
The code in `numpy/core/src/common/simd/neon` is the infrastructure used by [universal intrinsics](https://numpy.org/neps/nep-0038-SIMD-optimizations.html) to abstract away architecture-specific code inside the actual implementations. What is missing is that someone needs to rewrite the [avx512f code for minumum and maximum](https://github.com/numpy/numpy/blob/master/numpy/core/src/umath/simd.inc.src#L252) to use the universal intrinsics. xref gh-17985 which is doing the rewrite for add/subtract/multiply/divide

**Comment by neurolabusc** [comment]
Rosetta supports different forms of SSE, but [does not support any variant of AVX](https://developer.apple.com/documentation/apple_silicon/about_the_rosetta_translation_environment). The times for running each test on a macOS Intel i5-8259U are ~0.6s and on a Linux AMD 3900x ~0.33. Neither of those computers support avx512f.

Note that for many benchmarks the native code outperforms translated code. Here are the nibabel [benchmarks which rely on numpy](https://github.com/neurolabusc/AppleSiliconForNeuroimaging/blob/main/nibabel.tsv). The only test from that battery which showed a regression was bench_finite_range.py, which led me to drill down and determine that the np.min and np/max functions were specifically slow.

I also wrote a [C program](https://github.com/neurolabusc/simd/blob/master/main.cpp) that reports the maximum for an array of 64-bit doubles of the same size. The program times both scalar and SIMD (SSE/neon) instructions. The translated X86 code is much slower than native code for scalar, but they perform just as fast for the SIMD. The caveat with this test mimicking numpy's NaN propagation behavior (e.g. amax vs nanmax).

```
$g++ -O3 -o tstX86 main.cpp  -target x86_64-apple-macos10.12 -DmyDisableAVX; ./tstX86 4
Reporting minimum time for 4 tests
max64=1: min/mean	2091	2091	ms
max64SSE=1: min/mean	332	332	ms
max64(NaN)=1: min/mean	2086	2088	ms
max64SSE(NaN)=0.793516: min/mean	332	332	ms


$make -j; ./tst 4
g++ -O3 -o tst main.cpp -march=armv8-a+fp+simd+crypto+crc
Reporting minimum time for 4 tests
max64=1: min/mean	659	659	ms
max64SSE=1: min/mean	330	330	ms
max64(NaN)=1: min/mean	658	659	ms
max64SSE(NaN)=nan: min/mean	330	330	ms
```

**Comment by neurolabusc** [comment]
Just an update that same performance observed with numpy-1.20.1 with Python 3.9.2

**Comment by mattip** [comment]
Nothing will change until someone rewrites the code for max, min to use universal intrinsics.

**Comment by akbir** [comment]
Hi @mattip, I have access to hardware and would love to try get this moving. Can we make a specifc issue for this and I'll try get started? 

Will certainly need help but keen to sink some hours into this!

**Comment by neurolabusc** [comment]
@Developer-Ecosystem-Engineering numpy is a tremendously useful and popular tool.  Extending ARM support for universal intrinsics could have a profound impact.

**Comment by Developer-Ecosystem-Engineering** [comment]
> @Developer-Ecosystem-Engineering numpy is a tremendously useful and popular tool. Extending ARM support for universal intrinsics could have a profound impact.

Accelerate support was re-enabled here https://github.com/numpy/numpy/pull/18874, its worth checking with that support

**Comment by neurolabusc** [comment]
This fix was introduced in numpy 1.23 which was released on [June 22, 2022](https://pypi.org/project/numpy/#history). On the same computer as my original post:

```
                max all finite   0.14
                min all finite   0.14
                   max all nan   0.14
                   min all nan   0.14
python ./numpy_native_slower_than_translated.py   2.76s  user 0.07s system 295% cpu 0.959 total
```

Thanks!
